### PR TITLE
Use single DataBox type in DistributedObject and MockDistributedObject

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Replace the variant of DataBox types with a single DataBox type in (Mock)DistributedObject

Pros:
- Eliminates type aliases that determined the DataBox type for each action in the PDAL of a component
- Eliminates  code to visit the variants and find the correct DataBox type when executing
- No longer need simple actions and phase control to use Requires or if constexpr to determine if a tag is in the passed in DataBox and throw a runtime error if passed the wrong variant.  Removing the Requires or if constexpr will give a compiler error
- Phases and actions can be executed in any order without worrying about DataBox types

Cons:
- RemoveOptionsAndTerminatePhase no longer remove items from the DataBox; instead the items are reset via default construction.  If this leads to problems, we could convert these items to holding a std::optional that is set to nullopt when "removed".  

### Upgrade instructions

- No longer use Actions::SetupDataBox in a phase action list.  Setting up the DataBox is now handled by (Mock)DistributedObject
- ActionTesting::get_databox should only have a single template parameter (the component)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by kidder** [comment]
@markscheel In the last commit I had to change Test_InterpolatorReceivePoints.cpp so its mock action had the same arguments as the one it replaced.  Please look at

**Comment by kidder** [comment]
@knelli2 in the last commit I had to add tags to control_system::TestHelpers::MockObserverWriter.  Please look at that change and the other changes to the control system tests

File: src/Parallel/DistributedObject.hpp [hunk 1]
```diff
@@ -531,6 +493,11 @@ DistributedObject<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
             db::wrap_tags_in<Tags::FromGlobalCache, all_cache_tags>>>(
         metavariables{}, global_cache_proxy_,
         std::move(get<InitializationTags>(initialization_items))...);
+    box_ = db::create_from<
```
Comment (nilsdeppe) [inline src/Parallel/DistributedObject.hpp:496]: maybe for a future PR, but can we combine the `create_from` with the `create`? I don't remember what the most expensive part of DataBox construction is, but I thought `create_from` wasn't cheap at least

**Review by nilsdeppe — DISMISSED** [review]
LGTM, one general comment, but we should probably defer to another PR at some point

**Review by knelli2 — DISMISSED** [review]
Changes to the control system tests look good

File: src/Parallel/DistributedObject.hpp (hunk 1 above)
Comment (kidder) [reply src/Parallel/DistributedObject.hpp:496]: yes, but either need a helper function or a change to DataBox as currently can't mix initializing the mutable items with values passed in and default values

Comment (nilsdeppe) [reply src/Parallel/DistributedObject.hpp:496]: Ah, yea that's more difficult. Okay, definitely a different PR if we decide to do it at all :)

File: tests/Unit/Framework/ActionTesting.hpp [hunk 2]
```diff
@@ -423,28 +420,10 @@ struct InitializeDataBox<tmpl::list<SimpleTags...>, ComputeTagsList> {
...
-      const Parallel::GlobalCache<Metavariables>& /*cache*/,
-      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
-      const ParallelComponent* const /*meta*/) {
-    ERROR(
-        "Tried to apply ActionTesting::InitializeDataBox even though one or "
-        "more of its tags are already in the DataBox. Did you call next_action "
-        "too many times in the initialization phase?");
+    Initialization::mutate_assign<simple_tags>(
+        make_not_null(&box), std::forward<typename SimpleTags::type>(
+                                 tuples::get<SimpleTags>(initial_values_))...);
```
Comment (wthrowe) [inline tests/Unit/Framework/ActionTesting.hpp]: `move` instead of `forward`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Not for this PR, but we should also remove the DataBox return from actions, since it was only useful for changing the type.

**Comment by kidder** [comment]
> Not for this PR, but we should also remove the DataBox return from actions, since it was only useful for changing the type.

Yes, I was planning to do that in a followup PR, along with changing what is now the 2nd type in returned tuple to always be a AlgorithmExecution enum instead of sometimes that and sometimes a bool

**Review by wthrowe — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**